digitalocean_plugin/security.py

```diff
@@ -14,21 +14,138 @@
+        }
+
+    def delete_pubkey_from_account_by_fingerprint(self, fingerprint, **_):
+        """
+        :param fingerprint: key fingerprint
+        :param _: varargs
+        :return: True upon success
+        """
+        self._destroy_key(fingerprint)
+        return True
+
+    def _destroy_key(self, identifier):
+        url = self._build_url("account/keys/%s" % identifier)
+        h = self._common_headers()
+        response = requests.delete(url, headers=h)
+        code = response.status_code
+        if code != 204:
+            raise NonRecoverableError(
+                "Error on server. Expected status code = '204'"
+                ", but received '%s' instead." % str(code)
+            )
+        return
+
+    def delete_pubkey_from_account_by_keyid(self, keyid, **_):
+        """
+        :param keyid: id assigned to key (when imported) by digital ocean
+        :param _: varargs
+        :return: True upon success
+        """
+        self._destroy_key(keyid)
+        return True
+
+    def add_pubkey_to_digitalocean_account(self, pubkey_file, key_name, **_):
+        """
+        Uploads a public key file to the DigitalOcean account.
+        :param pubkey_file: full path to a public key file
+        :param name: an optional name for your key - will be assigned randomly
+         if not provided
+        :param _:
+        :return: The id and fingerprint of the uploaded key, returned as a
+        tuple (id, fingerprint)
+        """
+        if not os.path.isfile(pubkey_file):
+            raise NonRecoverableError("Unknown public key file: '{0}'."
+                                      .format(pubkey_file))
+        with open(pubkey_file, 'r') as f:
+            pubkey = f.read()
+        key_name = self._make_key_name(key_name)
+        assert pubkey, "A non-empty public key file must be provided."
+
+        payload = {
+            "name": key_name,
+            "public_key": pubkey
+        }
+
+        url = self._build_url('account/keys')
+        h = self._common_headers()
+
+        # try:
+        # because it's a POST, it should be only 1 response
+        # if we were to call GET, it would return all associated keys
+        response = requests.post(url, headers=h, data=payload)
+        code = response.status_code
+
+        if code < 200 or code > 299:
+            raise NonRecoverableError(
+                "Error on server for %(URL)s. Status code = '%(CODE)d'."
+                % {'URL': url, 'CODE': code}
+            )
+
+        r = response.json()['ssh_key']
+        return r['id'], r['fingerprint']
+
+    def _make_key_name(self, proposed_name):
+        """
+        :param proposed_name: a name to test for emptiness
+        :return: proposed_name, stripped (if non-empty), else a
+        randomly-generated name
+        """
+        if proposed_name:
+            return proposed_name.strip()
+        r = self.rand.randint(0, self.key_count_limit + 1)
+        return self.pubkey_stub.format(r)
```

Comment:
You might do something like this: https://github.com/cloudify-cosmo/cloudify-aws-plugin/blob/master/ec2/utils.py#L163

And you can also reuse it in tasks